File: src/commands/sync.ts
/** dot sync: pull, reconcile every tracked file with the host, commit, push. */

import { Task } from "../task.ts";
import {
  type ConfigError,
  configError,
  type GitError,
  type IoError,
  ioError,
} from "../errors.ts";
import { matchValue } from "../match.ts";
import { contractTarget, expandTarget } from "../path.ts";
import {
  hostLabel,
  type Layout,
  layout,
  loadManifest,
  loadState,
  saveState,
  type SyncState,
} from "../config.ts";
import { copyFile, readBytesIfExists, sha256, stat } from "../fs.ts";
import {
  commitIfChanged,
  git,
  gitInteractive,
  gitRaw,
  pushBestEffort,
} from "../git.ts";

type Facts = {
  readonly repoPath: string;
  readonly target: string;
  readonly hostPath: string;
  readonly hostHash: string | null;
  readonly repoHash: string | null;
  readonly baseHash: string | null;
};

/** What the three-way comparison detects for one tracked file. */
type Detected = "clean" | "toRepo" | "toHost" | "conflict" | "missing";

/**
 * How one tracked file was handled. A both-sides-changed file resolves into
 * "conflict" (host copy kept), "conflictRepo" (repo copy kept), "merged"
 * (a diff tool left both sides identical), or "skipped" (both sides left
 * untouched).
 */
type Plan = Detected | "conflictRepo" | "merged" | "skipped";

/**
 * Three-way comparison against the hash recorded at the last sync: an
 * unchanged side yields to the changed one; a change on both sides is a
 * conflict.
 */
const decide = (f: Facts): Detected => {
  if (f.repoHash === null && f.hostHash === null) return "missing";
  if (f.repoHash === null) return "toRepo";
  if (f.hostHash === null) return "toHost";
  if (f.repoHash === f.hostHash) return "clean";
  if (f.baseHash === f.repoHash) return "toRepo";
  if (f.baseHash === f.hostHash) return "toHost";
  return "conflict";
};

const hashOrNull = (
  bytes: Uint8Array | null,
): Task<string | null, IoError> =>
  bytes === null
    ? Task.of<string | null>(null)
    : sha256(bytes).map((h): string | null => h);

const gather = (
  l: Layout,
  state: SyncState,
  repoPath: string,
  target: string,
): Task<Facts, IoError> => {
  const hostPath = expandTarget(target, l.home);
  return readBytesIfExists(hostPath).andThen((host) =>
    readBytesIfExists(l.filesDir + "/" + repoPath).andThen((repo) =>
      hashOrNull(host).andThen((hostHash) =>
        hashOrNull(repo).map((repoHash): Facts => ({
          repoPath,
          target,
          hostPath,
          hostHash,
          repoHash,
          baseHash: state.files[repoPath] ?? null,
        }))
      )
    )
  );
};

type Outcome = {
  readonly plan: Plan;
  readonly repoPath: string;
  readonly target: string;
  /** Hash both sides hold after the action; null drops the state entry. */
  readonly hash: string | null;
};

const outcomeFor =
  (f: Facts) => (plan: Plan, hash: string | null): Outcome => ({
    plan,
    repoPath: f.repoPath,
    target: f.target,
    hash,
  });

const CHOICE_BY_INPUT = {
  l: "local",
  local: "local",
  r: "repo",
  repo: "repo",
  d: "tool",
  diff: "tool",
  tool: "tool",
  s: "skip",
  skip: "skip",
} as const;

type Choice = (typeof CHOICE_BY_INPUT)[keyof typeof CHOICE_BY_INPUT];

const CHOICES: Readonly<Record<string, Choice>> = CHOICE_BY_INPUT;

const menu = (target: string): string =>
  `conflict: ${target} changed both on this host and in the repo
  [l] keep local — the host copy wins; the repo copy stays in git history
  [r] keep repo  — overwrites the host copy
  [d] resolve in diff tool (git difftool)
  [s] skip — leave both sides as they are`;

/**
 * Reads one input line, one byte per read so nothing past the newline is
 * consumed — a later prompt still sees lines the user typed or pasted ahead.
 * Null at end of input.
 */
const readLine = (): string | null => {
  const buf = new Uint8Array(1);
  const bytes: number[] = [];
  for (;;) {
    const n = Deno.stdin.readSync(buf);
    if (n === null) {
      if (bytes.length === 0) return null;
      break;
    }
    const b = buf[0];
    if (n === 0 || b === undefined) continue;
    if (b === 10) break;
    if (b !== 13) bytes.push(b);
  }
  return new TextDecoder().decode(new Uint8Array(bytes));
};

/** Reads one resolution from the terminal; end of input counts as skip. */
const askChoice = (target: string): Task<Choice, IoError> =>
  Task.attempt(() => {
    console.log(menu(target));
    const ask = new TextEncoder().encode("choose [l/r/d/s]: ");
    for (;;) {
      Deno.stdout.writeSync(ask);
      const raw = readLine();
      if (raw === null) return Promise.resolve<Choice>("skip");
      const choice = CHOICES[raw.trim().toLowerCase()];
      if (choice !== undefined) return Promise.resolve(choice);
    }
  }, (u) => ioError("prompt", target, u));

/**
 * Opens git difftool on the host and repo copies; resolves to their common
 * hash when the tool leaves both sides identical, null otherwise.
 */
const mergeInTool = (
  f: Facts,
  repoFile: string,
): Task<string | null, GitError | IoError> =>
  gitInteractive(null, ["difftool", "-y", "--no-index", f.hostPath, repoFile])
    .andThen(() => readBytesIfExists(f.hostPath))
    .andThen((host) =>
      readBytesIfExists(repoFile).andThen(
        (repo): Task<string | null, IoError> =>
          host === null || repo === null
            ? Task.of<string | null>(null)
            : sha256(host).andThen((hostHash) =>
              sha256(repo).map((repoHash): string | null =>
                hostHash === repoHash ? hostHash : null
              )
            ),
      )
    );

/**
 * A both-sides-changed file: --force or a non-terminal stdin keeps the host
 * copy; otherwise the user picks the resolution per file. The host copy is
 * the only side git history cannot restore, so every path that discards it
 * is an explicit choice.
 */
const resolveConflict = (
  f: Facts,
  repoFile: string,
  force: boolean,
): Task<Outcome, IoError | GitError> => {
  const outcome = outcomeFor(f);
  const keepLocal = copyFile(f.hostPath, repoFile).map(() =>
    outcome("conflict", f.hostHash)
  );
  if (force || !Deno.stdin.isTerminal()) return keepLocal;
  return askChoice(f.target).andThen(
    (choice): Task<Outcome, IoError | GitError> =>
      matchValue(choice, {
        local: () => keepLocal,
        repo: () =>
          copyFile(repoFile, f.hostPath).map(() =>
            outcome("conflictRepo", f.repoHash)
          ),
        tool: () =>
          mergeInTool(f, repoFile).andThen(
            (merged): Task<Outcome, IoError | GitError> =>
              merged === null
                ? resolveConflict(f, repoFile, false)
                : Task.of<Outcome>(outcome("merged", merged)),
          ),
        skip: () => Task.of<Outcome>(outcome("skipped", f.baseHash)),
      }),
  );
};

const apply = (
  l: Layout,
  force: boolean,
  f: Facts,
): Task<Outcome, IoError | GitError> => {
  const repoFile = l.filesDir + "/" + f.repoPath;
  const outcome = outcomeFor(f);
  return matchValue(decide(f), {
    clean: () => Task.of<Outcome>(outcome("clean", f.hostHash)),
    missing: () => Task.of<Outcome>(outcome("missing", null)),
    toHost: () =>
      copyFile(repoFile, f.hostPath).map(() => outcome("toHost", f.repoHash)),
    toRepo: () =>
      copyFile(f.hostPath, repoFile).map(() => outcome("toRepo", f.hostHash)),
    conflict: () => resolveConflict(f, repoFile, force),
  });
};

const line = (o: Outcome, recover: string | null): string | null =>
  matchValue(o.plan, {
    clean: () => null,
    toRepo: () => `host -> repo  ${o.target}`,
    toHost: () => `repo -> host  ${o.target}`,
    conflict: () =>
      `host -> repo  ${o.target} (both sides changed: host copy kept)` +
      (recover === null ? "" : `\n  overwritten repo copy: ${recover}`),
    conflictRepo: () =>
      `repo -> host  ${o.target} (both sides changed: repo copy kept, host copy overwritten)`,
    merged: () =>
      `merged        ${o.target} (diff tool result kept on both sides)`,
    skipped: () =>
      `skipped       ${o.target} (conflict unresolved; rerun dot sync, or -f to keep the host copy)`,
    missing: () =>
      `missing       ${o.target} (gone on host and in repo; dot remove to untrack)`,
  });

/** Pulls with rebase; an empty remote (nothing pushed yet) counts as up to date. */
const pull = (l: Layout, branch: string): Task<boolean, GitError> =>
  gitRaw(l.repo, ["pull", "--rebase", "--autostash", "origin", branch])
    .andThen((out): Task<boolean, GitError> => {
      if (out.code === 0) return Task.of(true);
      if (out.stderr.includes("couldn't find remote ref")) {
        return Task.of(false);
      }
      return Task.fail({
        kind: "git",
        args: ["pull", "--rebase", "--autostash", "origin", branch],
        message: out.stderr.trim() || out.stdout.trim() ||
          `exit code ${out.code}`,
      });
    });

const requireBound = (l: Layout): Task<Layout, IoError | ConfigError> =>
  stat(l.repo + "/.git").andThen((info): Task<Layout, ConfigError> =>
    info === null
      ? Task.fail(configError("not bound — run: dot bind <repo>"))
      : git(l.repo, ["remote", "get-url", "origin"]).mapErr(() =>
        configError("no remote configured — run: dot bind <repo>")
      ).map(() => l)
  );

const summarize = (
  l: Layout,
  outcomes: readonly Outcome[],
  committed: boolean,
  pushWarning: string | null,
  preSync: string | null,
): string => {
  const repoDisplay = contractTarget(l.repo, l.home);
  const fileLines = outcomes.flatMap((o) => {
    const spec = `${preSync}:files/${o.repoPath}`;
    const recover = o.plan === "conflict" && preSync !== null
      ? `git -C ${repoDisplay} show ${/\s/.test(spec) ? `"${spec}"` : spec}`
      : null;
    const rendered = line(o, recover);
    return rendered === null ? [] : [rendered];
  });
  const clean = outcomes.filter((o) => o.plan === "clean").length;
  return [
    ...fileLines,
    ...(clean > 0 ? [`up to date: ${clean} file(s)`] : []),
    ...(outcomes.length === 0 ? ["nothing tracked — dot add <path>"] : []),
    ...(pushWarning !== null ? [pushWarning] : committed ? ["pushed"] : []),
  ].join("\n");
};

export const sync = (
  force: boolean,
): Task<string, ConfigError | IoError | GitError> =>
  Task.fromResult(layout()).andThen((l) =>
    requireBound(l)
      .andThen(() => git(l.repo, ["symbolic-ref", "--short", "HEAD"]))
      .andThen((branch) => pull(l, branch))
      .andThen(() => loadManifest(l))
      .andThen((manifest) =>
        loadState(l).andThen((state) => {
          const entries = Object.entries(manifest.files)
            .sort(([, a], [, b]) => a < b ? -1 : a > b ? 1 : 0);
          return Task.traverse(
            entries,
            ([repoPath, target]) =>
              gather(l, state, repoPath, target).andThen((f) =>
                apply(l, force, f)
              ),
          ).andThen((outcomes) => {
            const hashes = Object.fromEntries(
              outcomes.flatMap((o) =>
                o.hash === null ? [] : [[o.repoPath, o.hash] as const]
              ),
            );
            return saveState(l, { version: 1, files: hashes })
              .andThen(() =>
                commitIfChanged(l.repo, `dot: sync from ${hostLabel()}`)
              )
              .andThen((committed) => {
                // The sync commit's parent holds the repo copies that conflicts
                // overwrote; its hash pins the printed retrieval command.
                const preSync = committed &&
                    outcomes.some((o) => o.plan === "conflict")
                  ? git(l.repo, ["rev-parse", "--short", "HEAD^"])
                    .map((h): string | null => h)
                    .orElse(() => Task.of<string | null>(null))
                  : Task.of<string | null>(null);
                return preSync.andThen((ref) =>
                  pushBestEffort(l.repo).map((warning) =>
                    summarize(l, outcomes, committed, warning, ref)
                  )
                );
              });
          });
        })
      )
  );

File: src/commands/Sync.scala
package dot

import cats.effect.IO
import cats.syntax.all.*
import fs2.io.file.Path
import java.io.{FileDescriptor, FileInputStream}
import java.lang.ProcessBuilder.Redirect
import java.nio.charset.StandardCharsets.UTF_8
import mouse.all.*

/** dot sync: pull, reconcile every tracked file with the host, commit, push. */

/** What the three-way comparison detects for one tracked file. */
private enum Detected {

  /** Both copies hold the same content. */
  case Clean

  /** The host copy changed since the last sync, or the repo has no copy yet. */
  case ToRepo

  /** The repo copy changed since the last sync, or the host has no copy. */
  case ToHost

  /** Both copies changed since the last sync. */
  case Conflict

  /** Neither side has the file. */
  case Missing
}

/** How one tracked file was handled. */
private enum Plan {

  /** Nothing to do: both copies already match. */
  case Clean

  /** The host copy was copied into the repo. */
  case ToRepo

  /** The repo copy was installed on the host. */
  case ToHost

  /** Both sides had changed; the host copy was kept and overwrote the repo copy. */
  case Conflict

  /** Both sides had changed; the repo copy was kept and overwrote the host copy. */
  case ConflictRepo

  /** Both sides had changed; a conflict-marked copy waits under conflictsDir, both sides untouched. */
  case Parked

  /** A hand-edited parked copy was applied to both sides. */
  case Resolved

  /** Neither side has the file; the manifest entry is stale. */
  case Missing
}

/** How a both-sides-changed file is resolved. */
enum ConflictMode {

  /** Prompt per file when stdin is a terminal; keep the host copy otherwise. */
  case Ask

  /** Keep the host copy without asking. */
  case Force
}

/** What is known about one tracked file: both copies' hashes and the hash recorded at the last sync. */
private final case class Facts(
  repoPath: String,
  target: Target,
  hostPath: Path,
  hostHash: Option[String],
  repoHash: Option[String],
  baseHash: Option[String],
) {

  /**
   * Three-way comparison against the hash recorded at the last sync: an
   * unchanged side yields to the changed one; a change on both sides is a
   * conflict.
   */
  def decide: Detected = (repoHash, hostHash) match
    case (None, None)                               => Detected.Missing
    case (None, Some(_))                            => Detected.ToRepo
    case (Some(_), None)                            => Detected.ToHost
    case (Some(repo), Some(host)) if repo == host   => Detected.Clean
    case (Some(repo), _) if baseHash.contains(repo) => Detected.ToRepo
    case (_, Some(host)) if baseHash.contains(host) => Detected.ToHost
    case _                                          => Detected.Conflict

  def outcome(plan: Plan, hash: Option[String]): Outcome = Outcome(plan, repoPath, target, hash)
}

private object Facts {

  def gather(layout: Layout, state: SyncState, repoPath: String, target: Target): IO[Facts] = {
    val hostPath = target.expand(layout.home)
    for
      hostHash <- hostPath.sha256IfExists
      repoHash <- layout.repoFile(repoPath).sha256IfExists
    yield Facts(repoPath, target, hostPath, hostHash, repoHash, state.files.get(repoPath))
  }
}

private final case class Outcome(
  plan: Plan,
  repoPath: String,
  target: Target,
  /** Hash both sides hold after the action; None drops the state entry. */
  hash: Option[String],
) {

  /** The report line for this file; None when nothing changed. */
  def line(recover: Option[String], parkedAt: String): Option[String] = plan match
    case Plan.Clean    => None
    case Plan.ToRepo   => Some(s"host -> repo  $target")
    case Plan.ToHost   => Some(s"repo -> host  $target")
    case Plan.Conflict =>
      Some(
        s"host -> repo  $target (both sides changed: host copy kept)"
          + recover.map(r => s"\n  overwritten repo copy: $r").getOrElse(""),
      )
    case Plan.ConflictRepo =>
      Some(s"repo -> host  $target (both sides changed: repo copy kept, host copy overwritten)")
    case Plan.Parked =>
      Some(s"parked        $target (resolve $parkedAt, then dot sync; or dot sync --abort)")
    case Plan.Resolved =>
      Some(s"resolved      $target (parked copy applied to both sides)")
    case Plan.Missing =>
      Some(s"missing       $target (gone on host and in repo; dot remove to untrack)")
}

/** The process's standard input, read without the JVM's buffering so reads consume only what they return. */
private object Stdin {

  private val raw = FileInputStream(FileDescriptor.in)

  /**
   * Reads one line, one byte per read so nothing past the newline is consumed
   * — a later prompt still sees lines typed or pasted ahead. None at end of
   * input.
   */
  def readLine: Option[String] = {
    val bytes = scala.collection.mutable.ArrayBuffer.empty[Byte]
    var done = false
    var eof  = false
    while !done do
      val b = raw.read
      if b < 0 then
        if bytes.isEmpty then eof = true
        done = true
      else if b == 10 then
        done = true
      else if b != 13 then
        bytes += b.toByte
    if eof then None else Some(String(bytes.toArray, UTF_8))
  }

  /**
   * Whether the input is attached to a terminal, probed through a child
   * process that inherits the descriptor; an unprobeable platform falls back
   * to the JVM console check.
   */
  val isTerminal: IO[Boolean] = IO.blocking {
    try
      val pb = new ProcessBuilder("test", "-t", "0")
      pb.redirectInput(Redirect.INHERIT)
      pb.start.waitFor == 0
    catch
      case _: Exception => System.console != null
  }
}

/** The resolutions the conflict menu offers. */
private enum Choice {

  /** Keep the host copy; the repo copy stays retrievable from git history. */
  case Local

  /** Keep the repo copy; the host copy is overwritten. */
  case Repo

  /** Park a conflict-marked copy to resolve by hand; both sides stay. */
  case Skip
}

private object Choice {

  val byInput: Map[String, Choice] = Map(
    "l"     -> Local,
    "local" -> Local,
    "r"     -> Repo,
    "repo"  -> Repo,
    "s"     -> Skip,
    "skip"  -> Skip,
  )

  def menu(target: Target): String = {
    s"conflict: $target changed both on this host and in the repo\n"
      + "  [l] keep local — the host copy wins; the repo copy stays in git history\n"
      + "  [r] keep repo  — overwrites the host copy\n"
      + "  [s] skip — park a conflict-marked copy to resolve by hand; both sides stay"
  }

  /** Reads one resolution from the terminal; end of input counts as skip. */
  def ask(target: Target): IO[Choice] =
    IO.blocking {
      println(menu(target))
      var chosen: Option[Choice] = None
      var eof                    = false
      while chosen.isEmpty && !eof do
        print("choose [l/r/s]: ")
        System.out.flush
        Stdin.readLine match
          case None        => eof = true
          case Some(input) => chosen = byInput.get(input.trim.toLowerCase)
      chosen.getOrElse(Skip)
    }.orIoError("prompt", target.value)
}

extension (text: String) {
  private def hasConflictMarkers: Boolean = text.split("\n", -1).exists(_.startsWith("<<<<<<<"))
}

extension (src: Path) {

  /** Copies onto a host path; a permission failure names the sudo command that applies it by hand. */
  private def copyToHost(hostPath: Path): IO[Unit] =
    src.copyTo(hostPath).adaptError:
      case e: DotError.Io if e.cause == "permission denied" =>
        DotError.Io(e.op, e.path, s"permission denied — run: sudo cp $src $hostPath")
}

/** Everything constant across one sync run that reconciling a file needs. */
private final case class SyncCtx(layout: Layout, mode: ConflictMode, interactive: Boolean) {

  /**
   * Writes a conflict-marked merge of the host and repo copies (empty merge
   * base, so shared lines pass through and differing regions become marked
   * hunks) to the parked path. Both originals stay untouched.
   */
  def park(facts: Facts): IO[Unit] = {
    val base   = layout.root / "tmp-merge-base"
    val merged = Git.anywhere.mergeFile(
      ours = MergeSide(facts.hostPath, s"host: ${facts.target}"),
      base = MergeSide(base, "base: empty"),
      theirs = MergeSide(layout.repoFile(facts.repoPath), s"repo: files/${facts.repoPath}"),
    )
    for
      _    <- base.writeText("")
      text <- merged
      _    <- layout.parkedFile(facts.repoPath).writeText(text)
      _    <- base.removeIfExists
    yield ()
  }

  /**
   * A both-sides-changed file: Force (or a non-terminal stdin under Ask) keeps
   * the host copy, Ask lets the user pick the resolution per file. The host
   * copy is the only side git history cannot restore, so every path that
   * discards it is an explicit choice.
   */
  def resolveConflict(facts: Facts): IO[Outcome] = {
    val repoFile  = layout.repoFile(facts.repoPath)
    val keepLocal = facts.hostPath.copyTo(repoFile).as(facts.outcome(Plan.Conflict, facts.hostHash))
    mode match
      case ConflictMode.Force => keepLocal
      case ConflictMode.Ask   =>
        if !interactive then
          keepLocal
        else
          Choice.ask(facts.target).flatMap:
            case Choice.Local => keepLocal
            case Choice.Repo  =>
              repoFile.copyToHost(facts.hostPath).as(facts.outcome(Plan.ConflictRepo, facts.repoHash))
            case Choice.Skip =>
              park(facts).as(facts.outcome(Plan.Parked, facts.baseHash))
  }

  /**
   * A parked file overrides every mode: markers gone means the user resolved
   * it, so the content lands on both sides; markers still present hold the
   * conflict without re-asking. A parked file whose conflict no longer exists
   * is dropped.
   */
  def reconcile(facts: Facts): IO[Outcome] = {
    val repoFile   = layout.repoFile(facts.repoPath)
    val parkedFile = layout.parkedFile(facts.repoPath)
    def fresh: IO[Outcome] = facts.decide match
      case Detected.Clean    => IO.pure(facts.outcome(Plan.Clean, facts.hostHash))
      case Detected.Missing  => IO.pure(facts.outcome(Plan.Missing, None))
      case Detected.Conflict => resolveConflict(facts)
      case Detected.ToHost   =>
        repoFile.copyToHost(facts.hostPath).as(facts.outcome(Plan.ToHost, facts.repoHash))
      case Detected.ToRepo =>
        facts.hostPath.copyTo(repoFile).as(facts.outcome(Plan.ToRepo, facts.hostHash))
    parkedFile.readTextIfExists.flatMap:
      case None         => fresh
      case Some(parked) =>
        if !parked.hasConflictMarkers then
          for
            hash <- parkedFile.sha256IfExists

            _ <- parkedFile.copyToHost(facts.hostPath)
            _ <- parkedFile.copyTo(repoFile)
            _ <- parkedFile.removeIfExists
          yield facts.outcome(Plan.Resolved, hash)
        else if facts.decide == Detected.Conflict then
          IO.pure(facts.outcome(Plan.Parked, facts.baseHash))
        else
          parkedFile.removeIfExists *> fresh
  }
}

private def summarize(
  layout: Layout,
  outcomes: List[Outcome],
  pushed: Boolean,
  pushWarning: Option[String],
  preSync: Option[String],
): String = {
  val repoDisplay      = layout.display(layout.repo)
  val conflictsDisplay = layout.display(layout.conflictsDir)
  val fileLines        = outcomes.flatMap { outcome =>
    val recover = preSync.filter(_ => outcome.plan == Plan.Conflict).map { ref =>
      val spec   = s"$ref:files/${outcome.repoPath}"
      val quoted = if spec.exists(_.isWhitespace) then s"\"$spec\"" else spec
      s"git -C $repoDisplay show $quoted"
    }
    outcome.line(recover, conflictsDisplay + "/" + outcome.repoPath)
  }
  val clean          = outcomes.count(_.plan == Plan.Clean)
  val cleanLine      = Option.when(clean > 0)(s"up to date: $clean file(s)").toList
  val nothingTracked = outcomes.isEmpty.option("nothing tracked — dot add <path>").toList
  val pushLine       = pushWarning.orElse(pushed.option("pushed")).toList
  val lines          =
    fileLines
      ::: cleanLine
      ::: nothingTracked
      ::: pushLine
  lines.mkString("\n")
}

def sync(mode: ConflictMode): IO[String] = {
  for
    layout      <- Layout.resolve
    _           <- layout.requireBound
    interactive <- Stdin.isTerminal
    repo = Git.in(layout.repo)
    branch   <- repo.currentBranch
    _        <- repo.pull(branch)
    manifest <- Manifest.load(layout)
    state    <- SyncState.load(layout)

    ctx = SyncCtx(layout, mode, interactive)

    outcomes <- manifest.files.toList.sortBy(_._2).traverse: (repoPath, target) =>
      Facts.gather(layout, state, repoPath, target) >>= ctx.reconcile

    hashes = outcomes.flatMap: o =>
      o.hash.map(o.repoPath -> _)

    _         <- SyncState(hashes.toMap).save(layout)
    committed <- repo.commitIfChanged(s"dot: sync from $hostLabel")

    // The sync commit's parent holds the repo copies that conflicts
    // overwrote; its hash pins the printed retrieval command.
    preSync <-
      if committed && outcomes.exists(_.plan == Plan.Conflict)
      then repo.parentOfHead
      else IO.pure(None)

    // Push only when the remote lacks commits: a new one from this sync, or
    // one stranded by an earlier failed push.
    pending <-
      if committed
      then IO.pure(true)
      else repo.pendingPushes(branch).map(_ > 0)
    warning <-
      if pending
      then repo.pushBestEffort
      else IO.pure(None)
  yield summarize(layout, outcomes, pending && warning.isEmpty, warning, preSync)
}

/** dot sync --abort: discards every parked conflict; host and repo copies stay as they are. */
def syncAbort: IO[String] = {
  for
    layout <- Layout.resolve
    there  <- layout.conflictsDir.isPresent

    message <-
      if !there then
        IO.pure("no parked conflicts")
      else
        layout.conflictsDir.walkFiles.flatMap: files =>
          layout.conflictsDir.removeTreeIfExists.as:
            if files.isEmpty
            then "no parked conflicts"
            else s"discarded ${files.length} parked conflict(s)"
  yield message
}
